driver-core/src/main/com/mongodb/internal/logging/StructuredLogger.java

```diff
@@ -0,0 +1,210 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.logging;
+
+import com.mongodb.connection.ClusterId;
+import com.mongodb.internal.VisibleForTesting;
+import com.mongodb.internal.diagnostics.logging.Logger;
+import com.mongodb.internal.diagnostics.logging.Loggers;
+import com.mongodb.internal.logging.StructuredLogMessage.Entry;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.mongodb.internal.VisibleForTesting.AccessModifier.PRIVATE;
+import static java.lang.String.format;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public class StructuredLogger {
+
+    private final Logger logger;
+
+    private static final Map<String, StructuredLoggingInterceptor> INTERCEPTORS = new HashMap<>();
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    public static void addInterceptor(final String applicationName, final StructuredLoggingInterceptor interceptor) {
+        INTERCEPTORS.put(applicationName, interceptor);
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    public static void removeInterceptor(final String applicationName) {
+        INTERCEPTORS.remove(applicationName);
+    }
+
+    private static boolean hasInterceptor(final ClusterId clusterId) {
+        return INTERCEPTORS.containsKey(clusterId.getDescription());
+    }
+
+    private static StructuredLoggingInterceptor getInterceptor(final ClusterId clusterId) {
+        return INTERCEPTORS.get(clusterId.getDescription());
+    }
+
+    public StructuredLogger(final String suffix) {
+        this(Loggers.getLogger(suffix));
+    }
+
+    @VisibleForTesting(otherwise = PRIVATE)
+    public StructuredLogger(final Logger logger) {
+        this.logger = logger;
+    }
+
+    public boolean isDebugEnabled(final ClusterId clusterId) {
+        return logger.isDebugEnabled() || hasInterceptor(clusterId);
+    }
+
+    public void debug(final String message, final ClusterId clusterId, final String format,
+            final String k1, final Object v1,
+            final String k2, final Object v2,
+            final String k3, final Object v3,
+            final String k4, final Object v4,
+            final String k5, final Object v5,
+            final String k6, final Object v6) {
+        if (hasInterceptor(clusterId)) {
+            getInterceptor(clusterId).intercept(new StructuredLogMessage(logger.getName(), "debug", message, clusterId,
+                    new Entry(k1, v1), new Entry(k2, v2), new Entry(k3, v3), new Entry(k4, v4), new Entry(k5, v5), new Entry(k6, v6)));
+        }
+        if (logger.isDebugEnabled()) {
+            logger.debug(format(format, v1, v2, v3, v4, v5, v6));
+        }
+    }
+
+    public void debug(final String message, final ClusterId clusterId, final String format,
```

Comment:
Some of these methods aren't used, so should be removed?

An alternative API might be

```
debug(final String message, final ClusterId clusterId, final Throwable exception, final String format, final Object... arguments)
```

Follows the slf4j convention and then you can process the arguments as kv pairs.

But I suppose that adds overhead.

Reply:
Removed unused methods.

Yes, it would add overhead and I also appreciate the type safety that you get with the current approach